src/bespokelabs/curator/request_processor/online_request_processor.py

```diff
@@ -0,0 +1,312 @@
+
+@dataclass
+class StatusTracker:
+    """Tracks the status of all requests."""
+
+    num_tasks_started: int = 0
+    num_tasks_in_progress: int = 0
+    num_tasks_succeeded: int = 0
+    num_tasks_failed: int = 0
+    num_tasks_already_completed: int = 0
+    num_api_errors: int = 0
+    num_other_errors: int = 0
+    available_request_capacity: float = 0
+    available_token_capacity: float = 0
+    last_update_time: float = field(default_factory=time.time)
+    max_requests_per_minute: int = 0
+    max_tokens_per_minute: int = 0
+    pbar: tqdm = field(default=None)
+    response_cost: float = 0
+
+    def __str__(self):
+        return (
+            f"Tasks - Started: {self.num_tasks_started}, "
+            f"In Progress: {self.num_tasks_in_progress}, "
+            f"Succeeded: {self.num_tasks_succeeded}, "
+            f"Failed: {self.num_tasks_failed}, "
+            f"Already Completed: {self.num_tasks_already_completed}\n"
+            f"Errors: {self.num_other_errors}"
+        )
+
+    def update_capacity(self):
+        """Update available capacity based on time elapsed"""
+        current_time = time.time()
+        seconds_since_update = current_time - self.last_update_time
+
+        self.available_request_capacity = min(
+            self.available_request_capacity
+            + self.max_requests_per_minute * seconds_since_update / 60.0,
+            self.max_requests_per_minute,
+        )
+
+        self.available_token_capacity = min(
+            self.available_token_capacity
+            + self.max_tokens_per_minute * seconds_since_update / 60.0,
+            self.max_tokens_per_minute,
+        )
+
+        self.last_update_time = current_time
+
+    def has_capacity(self, token_estimate: int) -> bool:
+        """Check if there's enough capacity for a request"""
+        self.update_capacity()
+        return (
+            self.available_request_capacity >= 1 and self.available_token_capacity >= token_estimate
+        )
+
+    def consume_capacity(self, token_estimate: int):
+        """Consume capacity for a request"""
+        self.available_request_capacity -= 1
+        self.available_token_capacity -= token_estimate
+
+
+@dataclass
+class APIRequest:
+    """Stores an API request's inputs, outputs, and other metadata."""
+
+    task_id: int
+    generic_request: GenericRequest
+    api_specific_request: dict
+    attempts_left: int
+    result: list = field(default_factory=list)
+    prompt_formatter: PromptFormatter = field(default=None)
+    created_at: datetime.datetime = field(default_factory=datetime.datetime.now)
+
+
+class OnlineRequestProcessor(BaseRequestProcessor, ABC):
+    """Abstract base class for online request processors that make real-time API calls."""
+
+    def __init__(
+        self,
+        model: str,
+        temperature: Optional[float] = None,
+        top_p: Optional[float] = None,
+        presence_penalty: Optional[float] = None,
+        frequency_penalty: Optional[float] = None,
```

Comment:
generation config, support all litellm params?

Reply:
let's revisit this when working on #62